common.mill

```diff
@@ -0,0 +1,303 @@
+    "cat=deprecation&origin=chisel3\\.experimental\\.IntrinsicModule:s",
+    "cat=deprecation&origin=chisel3\\.ltl.*:s",
+    // Deprecated for external users, will eventually be removed.
+    "cat=deprecation&msg=Looking up Modules is deprecated:s",
+    // Only for testing of deprecated APIs
+    "cat=deprecation&msg=Use of @instantiable on user-defined types is deprecated:s"
+  )
+
+  // ScalacOptions
+  val scala2CommonOptions = Seq(
+    "-deprecation",
+    "-feature",
+    "-unchecked",
+    "-Werror",
+    "-Ymacro-annotations",
+    "-explaintypes",
+    "-Xcheckinit",
+    "-Xlint:infer-any",
+    "-Xlint:missing-interpolator",
+    "-language:reflectiveCalls",
+    s"-Wconf:${scala2WarnConf.mkString(",")}"
+  )
+
+
+trait HasScala2MacroAnno extends CrossModuleBase {
+  override def scalacOptions = T {
+    if (!isScala3(crossScalaVersion)) {
+      super.scalacOptions() ++ Agg("-Ymacro-annotations")
+    } else super.scalacOptions()
+  }
+}
+
+trait HasScala2Plugin extends CrossModuleBase {
+  def pluginModule: PluginModule
+
+  override def scalacOptions = T {
+    if (!isScala3(crossScalaVersion)) {
+      super.scalacOptions() ++ Agg(s"-Xplugin:${pluginModule.jar().path}")
+    } else super.scalacOptions()
+  }
+
+  override def scalacPluginClasspath = T {
+    if (!isScala3(crossScalaVersion)) {
+      super.scalacPluginClasspath() ++ Agg(pluginModule.jar())
+    } else super.scalacPluginClasspath()
+  }
+}
+
+trait ChiselPublishModule extends CiReleaseModule {
+  // Publish information
+  def pomSettings = PomSettings(
+    description = artifactName(),
+    organization = "org.chipsalliance",
+    url = "https://www.chisel-lang.org",
+    licenses = Seq(License.`Apache-2.0`),
+    versionControl = VersionControl.github("chipsalliance", "chisel"),
+    developers = Seq(
+      Developer("jackkoenig", "Jack Koenig", "https://github.com/jackkoenig"),
+      Developer("azidar", "Adam Izraelevitz", "https://github.com/azidar"),
+      Developer("seldridge", "Schuyler Eldridge", "https://github.com/seldridge")
+    )
+  )
+
+  override def sonatypeHost = Some(SonatypeHost.s01)
+
+  override def publishVersion = VcsVersion
+    .vcsState()
+    .format(
+      countSep = "+",
+      revHashDigits = 8,
+      untaggedSuffix = "-SNAPSHOT"
+    )
+
+}
+
+trait FirrtlModule extends CrossSbtModule with Cross.Module[String] with HasScala2MacroAnno with ScalafmtModule {
+  def scalaVersion = crossScalaVersion
+
+  override def scalacOptions = T {
+    if (isScala3(crossScalaVersion)) {
+      Seq.empty[String]
+    } else {
+      scala2CommonOptions ++ Seq(
+        "-language:reflectiveCalls",
+        "-language:existentials",
+        "-language:implicitConversions",
+        "-Yrangepos", // required by SemanticDB compiler plugin
+        "-Xsource:3",
+        "-Xsource-features:infer-override"
+      )
+    }
+  }
+  
+  def scoptIvy: Dep
+  def commonTextIvy: Dep
+  def osLibIvy: Dep
+  def json4sIvy: Dep
+  def dataclassIvy: Dep
+
```

Comment:
I do not understand the purpose of this sort of software design for the build, why do we need `trait FirrtlModule`. Why do we have traits with abstract methods when we can easily just define our ivyDeps concretely and be done with it?

Reply:
+1 neither do I agree to leave a common file.

But I agree we should bump to latest mill build system.

You may need to split into multiple mill package into multiple folder to keep the build.mill tidy to follow the up to date mill build policy

Reply:
Agree with upgrading Mill.

@unlsycn (or anyone else looking into this), make sure to look at https://github.com/com-lihaoyi/mill/discussions/3840 as we will need to add a `.mill-jvm-opts` that sets `chisel.project.root=${PWD}` to ensure the source locators work correctly.

Reply:
> I do not understand the purpose of this sort of software design for the build, why do we need `trait FirrtlModule`. Why do we have traits with abstract methods when we can easily just define our ivyDeps concretely and be done with it?

You are right, it is originally for reusing the mill scripts in other repositories that treat Chisel as a dependency. I just realized that this is unnecessary for our flow (so I marked this PR as draft😂) and consider splitting it into mutiple package.mill in subfolders.